src/train_cmd.cpp

```diff
@@ -2160,7 +2160,17 @@ CommandCost CmdForceTrainProceed(DoCommandFlags flags, VehicleID veh_id)
 		 * to proceed to the next signal. In the other cases we
 		 * would like to pass the signal at danger and run till the
 		 * next signal we encounter. */
-		t->force_proceed = t->force_proceed == TFP_SIGNAL ? TFP_NONE : t->flags.Test(VehicleRailFlag::Stuck) || t->IsChainInDepot() ? TFP_STUCK : TFP_SIGNAL;
+		if (t->force_proceed == TFP_SIGNAL) {
+			t->force_proceed = TFP_NONE;
+		} else if (t->flags.Test(VehicleRailFlag::Stuck)) {
+			TileIndex next_tile = TileAddByDiagDir(t->tile, TrackdirToExitdir(t->GetVehicleTrackdir()));
+			if (next_tile == INVALID_TILE) t->force_proceed = TFP_STUCK;
+			else {
```

Comment:
Coding style: if using an else, then you may not use the `if` without brackets.
```suggestion
			if (next_tile == INVALID_TILE) {
				t->force_proceed = TFP_STUCK;
			} else {
```

Alternatively, if you move this code to a separate function, you can reduce the number of brackets significantly and eliminate the else branches. If you choose to go this path, please do incorporate the preceding comments in the Doxygen-documentation of the function.
```
static TrainForceProceeding DetermineNextTrainForceProceeding(const Train *t)
{
    if (t->force_proceed == TFP_SIGNAL) return TFP_NONE;
    if (!t->flags.Test(VehicleRailFlag::Stuck)) return t->IsChainInDepot() ? TFP_STUCK : TFP_SIGNAL;
    
    TileIndex next_tile = TileAddByDiagDir(t->tile, TrackdirToExitdir(t->GetVehicleTrackdir()));
    if (next_tile == INVALID_TILE) return t->force_proceed = TFP_STUCK;
    return IsTileType(next_tile, MP_RAILWAY) && HasSignals(next_tile) ? TFP_SIGNAL : TFP_STUCK;
}
```

Reply:
I went for the helper function, because in my opinion it is more readable.